File: dart/packages/aoc2015/lib/src/day03.dart
// https://adventofcode.com/2015/day/3

import 'package:aoc/aoc.dart';
import 'package:aoc/util/string.dart';
import 'package:aoc/util/vec.dart';

main() => Day03().solve();

class Day03 extends AdventDay {
  Day03() : super(2015, 3);

  @override
  dynamic part1(String input) {
    Vec current = Vec.zero;
    final housesVisited = <Vec>{current};
    for (final dir in input.chars.map((c) => _directionOffset[c]!)) {
      current += dir;
      housesVisited.add(current);
    }
    return housesVisited.length;
  }

  @override
  dynamic part2(String input) {
    Vec santa = Vec.zero;
    Vec robotSanta = Vec.zero;
    bool santaTurn = true;
    final housesVisited = <Vec>{santa};

    for (final dir in input.chars.map((c) => _directionOffset[c]!)) {
      if (santaTurn) {
        santa += dir;
        housesVisited.add(santa);
      } else {
        robotSanta += dir;
        housesVisited.add(robotSanta);
      }
      santaTurn = !santaTurn;
    }
    return housesVisited.length;
  }

  static const _directionOffset = <String, Vec>{
    '^': Vec.up,
    '>': Vec.right,
    'v': Vec.down,
    '<': Vec.left,
  };
}

File: clojure/src/aoc2017/day03.clj
;; https://adventofcode.com/2017/day/3
(ns aoc2017.day03
  (:require
   [aoc.day :as d]
   [aoc.util.math :as m]
   [aoc.util.pos :as p]
   [aoc.util.string :as s]))

(defn input [] (d/day-input 2017 3))

(defn- spiral-dirs []
  (let [ring (mapcat #(repeat 2 %) (map inc (range)))
        dir (cycle [p/dir-right p/dir-up p/dir-left p/dir-down])]
    (mapcat repeat ring dir)))

(defn- spiral-positions []
  (reductions p/pos+ p/origin (spiral-dirs)))

(defn- sum-adjacent [grid pos]
  (reduce + (keep grid (p/adjacent-to pos))))

(defn- first-greater [target]
  (reduce
   (fn [grid pos]
     (let [value (sum-adjacent grid pos)]
       (if (> value target)
         (reduced value)
         (assoc grid pos value))))
   {p/origin 1}
   (rest (spiral-positions))))

(defn part1 [input]
  (->> (s/int input)
       dec
       (nth (spiral-positions))
       (m/manhattan-distance p/origin)))

(defn part2 [input] (first-greater (s/int input)))
